Significant overhaul of the OpenAL implementation

jpernst opened the pull request:
This PR represents substantial work on improving the OpenAL implementation by @yoanlcq and myself.

First and foremost, every entry point defined by the OpenAL 1.1 specification is implemented, which is important for bindings implementations that expose the full API.

Second, error checking is now quite rigorous and conformant with the spec, which should help prevent incompatibilities with native implementations. In general, the semantics of the API conform to the spec whenever possible, including seemingly useless cases such as 0-length buffers.

Third, missing functionalities such as Doppler shift and listener-relative sources are restored.

Finally, several common extensions are implemented, including:
- "ALC_SOFT_pause_device"
- "ALC_SOFT_HRTF"
- "AL_EXT_float32"
- "AL_SOFT_loop_points"
- "AL_SOFT_source_length"
- "AL_EXT_source_distance_model"
- "AL_SOFT_source_spatialize"

I have plans to implement at least some of the EFX extension, but that one is massive, so I'm leaving it for a later PR.

Several new tests have been added to help ensure the implementation is correct, but OpenAL has a fairly large surface area, so exposing it to a few non-trivial apps to really shake it out first is probably a good idea as well.

Very little of the original implementation remains, so I understand completely if this is a controversial or time consuming PR to examine, but I hope it's found useful. I've tried to "do as the GL bindings do" for the GetProcAddress functions and such, but it's entirely possible that it will need adjustment.

kripken commented on tools/system_libs.py:
```diff
@@ -246,6 +247,12 @@ def create_gl(libname): # libname is ignored, this is just one .o file
+  # al
+  def create_al(libname): # libname is ignored, this is just one .o file
+    o = in_temp('al.o')
+    check_call([shared.PYTHON, shared.EMCC, shared.path_from_root('system', 'lib', 'al.c'), '-o', o])
```
Perhaps this should be built with `-O2` or `-Os`?

jpernst replied:
I wasn't sure what the best choice was, so I just made it exactly the same as the `create_gl` version. The functions contained inside are generally only called a few times at the beginning of a program and never again, so perhaps `-Os` would be best.

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+Audio
+=====
+
+Emscripten supports OpenAL 1.1 "out of the box", with no additional compilation flag required. The implementation uses the Web Audio API.
```
While this has been true for OpenAL, we actually want to move away from this model and use the same linker directive that native systems use to link to OpenAL. That is, use `-lopenal` to link to OpenAL which I believe Linux systems do(?) The PR #4665 made this directive available, with the intent of making that mandatory in the future. It would be good to document that possibility here as the "desired" mechanism.

yoanlcq replied:
Will fix! Also yes, `-lopenal` is the Linux way.

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+Guidelines for Audio on Emscripten
+==================================
+
+First: Do not loop on OpenAL calls.
```
This would be good to be clarified to refer to synchronously blocking on long running calls to be more clear. (accessing OpenAL calls in a loop is fine, one just can't block for long periods)

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+What you must do instead is perform each such query only once per "main loop iteration" (i.e the callback you provide via :c:func:`emscripten_set_main_loop` or :c:func:`emscripten_set_main_loop_arg`).
+
+
+Second: Avoid creating and destroying resources relentlessly.
```
This second point would probably be good to be removed altogether. Referencing poor garbage collection implementations in browsers can read a bit FUD-mongering. In general we do not find OpenAL garbage collection to be a big performance problem, but `alBufferData` is the problematic one. Referring to the usual "do resource loading in cold paths" optimization would be ok however - that is also advice that is sound for native code.

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. note:: 
+    On Firefox, the user is also given the opportunity to pick from a list of devices at that moment.
+    On other browsers, this is normally a simple "Allow/Deny" button pair on a toast.
```
This browser specific note is a candidate to get out of date on the long run, and does not seem to carry particularly important value, so would be cleaner to avoid this? (or perhaps remark above `asking for the user's permission, and on some browsers, the capture device to choose)`.

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+If the user clicks "Deny", the device is invalidated (because this is somewhat
+similar to unplugging the physical device) and calls to ``alcCapture*`` functions on that
+device then consistently fail with ``ALC_INVALID_DEVICE``.
+Your application must be prepared to handle this properly.
```
Change `must` to `should`?

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+===============================================
+
+Internally, Web Audio's capture data is always backed by a Javascript ``Float32Array``.  
+Thus, ``AL_FORMAT_MONO_FLOAT32`` and ``AL_FORMAT_STEREO_FLOAT32`` are the only formats which do not require running "type conversion passes" on acquired buffers.
```
Does `"type conversion passes"` need double quotes?

yoanlcq replied:
I had put them because it felt like this was a term I just pulled of nowhere and needed to show that I was aware of it. It's hard to put into words, but not important anymore since I'll change the wording to be more to the point, like : 
```
Thus, ``AL_FORMAT_MONO_FLOAT32`` and ``AL_FORMAT_STEREO_FLOAT32`` 
are the only formats which do not require converting acquired samples from their
initial type to another.
```

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. note::
+    Some browsers "remember" this choice and apply it automatically every time it would be asked again instead.
+    There's no way for the implementation to detect this behavior.
```
Good note!

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+- It's probably pointless to pick a stereo format for capture;
+- float32 samples use more memory than the standard 16-bit or 8-bit integer formats;
+- Most microphones aren't very high quality.
+  Especially for real-time apps such as games, users will likely tolerate lower quality recordings.
```
Let's remove all these four notes, they don't seem to carry particular value.

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+- Most microphones aren't very high quality.
+  Especially for real-time apps such as games, users will likely tolerate lower quality recordings.
+
+Also, the actual sample rate at which samples are acquired from the device is unpredictable and depends on many factors, including the user's browser and setup. If this sample rate does not match the one your app requests, the implementation is required to perform resampling on your behalf.  
```
Change `is unpredictable` to something like `is currently dictated by the browser and hardware, instead of user code`, since `unpredictable` might read "random", but the behavior is very specific and deterministic on a specific browser and OS. See also #5349, which is a new update to Web Audio spec, though not sure if it will apply to microphone captures.

yoanlcq replied:
Agreed for the rewording.
Also, interesting! Looks like something worth adding to the `Improving` section at the end.

juj commented on site/source/docs/porting/Audio.rst:
```diff
@@ -0,0 +1,143 @@
+Also, the actual sample rate at which samples are acquired from the device is unpredictable and depends on many factors, including the user's browser and setup. If this sample rate does not match the one your app requests, the implementation is required to perform resampling on your behalf.  
+
+That sample rate is given by ``audioCtx.sampleRate``, where ``audioCtx`` is the ``AudioContext`` object used internally by the relevant capture ``ALCdevice``.
+Currently, Emscripten provides no direct way for applications to access this value, but this might be provided through an Emscripten-specific OpenAL extension (which is not here yet because it requires registration).
```
Does this reference to asking for the actual sampling rate of an already created context, or the default sampling rate of created contexts? Adding functions for either as needed sounds good.

yoanlcq replied:
`That sample rate [...]` refers to the sampling rate of already created contexts. However I'm a bit confused by the question because it's not clear what we mean by "context" (which I assume is `AudioContext`).
The thing is, it just so happens that the sample rate is the same across all `AudioContext` on the same setup (browser, OS, hardware, etc). This is not guaranteed anywhere, it's just a (naïve?) observation which I still wanted to document as such.

> Adding functions for either as needed sounds good.

I believe the "standard" way would be to register a constant for use by `alcGetIntegerv()`. Adding new dedicated functions which directly return a single int would also be fine I guess, but I can't decide which way is better.  
The reason I (we?) didn't bother registering this functionality as an extension _yet_ is that this is an implementation detail not _that_ many apps actually care about, because AFAIK it's not provided in native OpenAL impls anyway.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4542 @@
+
+    sourceDuration: function(src) {
+      var length = 0.0;
+      for (i = 0; i < src.bufQueue.length; i++) {
```
In these types of loops, use `for (var i = 0; ...)`

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4542 @@
+        }
+
+        listener._direction = value.slice(0, 3);
+        listener._up = value.slice(3, 6);
```
This looks like a source of per-frame generated garbage. Not a big deal, though it would be good to avoid temporary garbage where possible.

jpernst replied:
Removed the slicing here since the destination arrays are already guaranteed to exist.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4542 @@
+    case 0x1101:  /* AL_FORMAT_MONO16 */
+    case 0x1100:  /* AL_FORMAT_MONO8 */
+      outputChannelCount = 1;
+
```
The kind of generic sample format conversion mechanism below looks clean, though these kinds of loops have been profiled to be the biggest performance bottleneck in audio in the past. See https://github.com/kripken/emscripten/blob/incoming/src/library_openal.js#L857 for how these were optimized down to the tightest form or JS loops for best performance. Not sure if the performance issue might carry over to microphone, but might be good to keep in mind.

juj replied:
Convention is to have a space after if: `if (`.

juj replied:
ScriptProcessorNode is being deprecated in Web Audio, so it would be good to not rely on this (https://www.w3.org/TR/webaudio/#the-scriptprocessornode-interface---deprecated). Is there other ways to implement this, or do you think this will need upcoming AudioWorklets?

juj replied:
Space after for, `for (`

juj replied:
In general, `for(var i = 0; I < srcArray.length; ++i) {`

juj replied:
Either

```javascript
if (condition) statement;
```

or

```javascript
if (condition) {
  statement;
}
```

juj replied:
Let's remove the word `helper` from all these functions, it's a bit redundant, and just have `AL.getGlobal` et al.?, (or something like `AL.getGlobalParameter` and `AL.setGlobalParameter`?)

juj replied:
This kind of pattern is explicitly reverting this earlier PR: https://github.com/kripken/emscripten/pull/4275/files#diff-36fcd036fa88446c972f56ca850c357fL1387. I see your documentation did have a mention about being sensitive to garbage collection pressure, so it would be good to retain the garbage free operation from PR #4275?

juj replied:
`AL.setListenerParameter` instead of `AL.listenerHelper`? The word helper reads a bit ambiguous about what it is exactly helping.

juj replied:
With this pattern I see that there's implicit casting that occurs to convert the parameter type to the type that was requested. In https://www.openal.org/documentation/OpenAL_Programmers_Guide.pdf pages 12&13, I understand it is not allowed to query an integer parameter via a float get function, and vice versa. Does that seem correct to you?

yoanlcq replied:
I was actually asked the same question before [(here)](https://github.com/jpernst/emscripten/issues/2#issuecomment-312729735), and basically according to [the MDN doc on `ScriptProcessorNode`](https://developer.mozilla.org/en/docs/Web/API/ScriptProcessorNode) the functionality is to be replaced by Audio Workers.  
Following [the link in the deprecation message](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API#Audio_Workers), I saw that the latter was not implemented yet (even though it dates back to 2014), so I didn't go any further, not even considering opening the spec to know more. (not claiming it was the right thing to do, it's just how it happened)

I also believe that `ScriptProcessorNode` won't be an easy one to bury.

Also, AudioWorklets, Web Workers, etc. are more involved and quite another plane in our problem-space if I may say.

yoanlcq replied:
I'll look into it. I'm a bit afraid this might end up less easy to read, but at the same time it doesn't look hard to do and the benefits are clear.

yoanlcq replied:
(If we do it, I'd be in favor in`getGlobal`, it's short and no less clear. @jpernst ?)

jpernst replied:
i think `getGlobalParam`/etc is a nice compromise between brevity and clarity

yoanlcq replied:
(I didn't work on this part of the code, just wanted to add that the spec seems to agree, and it's probably expected to fail with `ALC_INVALID_ENUM` in this case.)

yoanlcq replied:
Also agreed, waiting for @jpernst's opinion on this

jpernst replied:
Yeah, it was probably a copy/pase. I'll tackle this and the other main AL semantic issues.

jpernst replied:
Same. I think `Param` would do fine, but if we want the full word that's OK too.

jpernst replied:
Actually, upon closer review, the 1.1 specification does expilcitly allow int/float queries for AL_POSITION and such, while the Programming Guide does not. I chose to adhere to the spec document in cases where they contradict.

jpernst replied:
Passing an array is very useful to maintain the separation of the core param setter methods and the entrypoints, so what I've done is create a recycled param array that will be filled before calls to the real setters, then passed in, This should hopefully have the desired effect of eliminating that source of per-frame garbage.

yoanlcq replied:
Fixed (?) by [commit e344c](https://github.com/kripken/emscripten/pull/5367/commits/e344c352b5cfd69f2109f8d80f0e2efd752a7708).  

Should I also replace the `setSample()` mechanism in `alcCaptureSamples()` to use heap views instead, and unroll loops the same way ?  
I personally would rather avoid this task because, since we handle the "resampling" and "no resampling" cases separately, it would cause a local explosion of code, which would harm readability (and maintenance) really bad this time - all for the name of better performance, but do we need to go that far? Aren't today's JS engines good enough for this ? (genuine question)

IMO it would be best to leave it as-is (or close to "as-is") for the time being, and if it (`alcCaptureSamples()`) actually shows up in profilers, people can always improve it after our PR (this time with real-world performance data in their hands).  
As I see it right now, the added burden simply outweights the hypothetical benefits.

juj commented:
> Very little of the original implementation remains, so I understand completely if this is a controversial or time consuming PR to examine, but I hope it's found useful.

Thanks. This PR is appreciated, but agreed that it is quite difficult to review, because there are so many conceptually unrelated changes that the diffs no longer line up on functions even. Added some superficial comments, I'll schedule time to review this closer with tests shortly.

yoanlcq commented:
@juj, starting from your first review on line 1605 at `src/library_openal.js`, all diffs in your comments show up as going from lines 1605 to 1608, so I (we?) can't tell precisely which part of the code your comments actually refer to. I suspect (and hope) that I'm the only one experiencing this, but this is how it appears both in my mails and on GitHub, so there's something weird going on.  

Right now I'm off to make changes to the docs and fixing style issues as per your comments.

jpernst commented:
I've rebased on incoming again to eliminate the merge commits since the emscripten contributor guide says to avoid those.

yoanlcq commented:
Not wanting to be an annoyance, but are there any plans for carrying this PR forward @juj ? I just fear this PR might "sink", and without updates it's hard to tell. We can wait for as long as needed of course, but it would be nice to have some sort of ETA if you have any (if you don't, it's fine too).

juj commented:
> Not wanting to be an annoyance, but are there any plans for carrying this PR forward @juj ? I just fear this PR might "sink", and without updates it's hard to tell. We can wait for as long as needed of course, but it would be nice to have some sort of ETA if you have any (if you don't, it's fine too).

Thanks for the ping and sorry for the delay, I was on a 2 week PTO and back today. This PR does look good and I think we do want to work this in to merge, I'll give your improvements a second review pass as soon as I get my email queue cleared.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4591 @@
           }
 #if OPENAL_DEBUG
           else {
-            Runtime.warnOnce('Unable to start AudioBufferSourceNode playback! Not supported by the browser?');
```
This PR changes all instances of strings from single quotes `'` to double quotes `"`. The convention in this repository in JS code is to use single quotes - is there a specific motivation for this change?

jpernst replied:
Just force of habit; I've changed them all back single quotes.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4591 @@
+      } else {
+        var skipCount = 0;
+        // In the absence of scheduled buffers, fast-forward playback state to the current time
+        while (true) {
```
This `while(true)` is making me nervous because after staring at the loop body for 15 minutes, I cannot easily statically conclude that the loop will terminate in all possible cases. Could this be a `for(var i = 0; i < src.bufQueue.length; ++i)` loop instead? If I understand correctly, we only need to iterate as many times as there are queued buffers for a source?

jpernst replied:
The purpose of this loop is bring the openal impl's internal state back up to date to match the current time of the webaudio context. It's designed to handle arbitrarily long pauses that could be caused by any kind of blocking delay elsewhere in the code. If a looping buffer queue was started prior to a long pause, it could have looped any number of times in the interim, so we need to loop as many times as it takes to get to the correct playback states.

The primary termination condition is when the next computed buffer start time is in the future, which is calculated by repeatedly adding buffer durations, so the value should be monotonically increasing. A queue full of 0-length buffers can't cause an infinite loop since those are explicitly checked for and the loop aborts if the number of 0-length buffers (skipCount) becomes equal to the queue length.

Still, we could add some kind of emergency abort threshold just in case, if you feel it's necessary.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4591 @@
+      // WebAudio expects world space coordinates, so if the source is listener-relative
+      // we must transform the coordinates from listener space into world space.
+      if (src.relative) {
+        // Negate the listener direction since forward is -Z.
```
This comment is somewhat ambiguous. Here it would be good to document what the different coordinate spaces are for Web Audio and OpenAL, to illustrate what the needed coordinate transform that is performed here is.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4591 @@
+        // Translate the source position by the listener position to finish.
+        posX += lPosX;
+        posY += lPosY;
+        posZ += lPosZ;
```
This math does not look correct according to the comment above, `lPos` is in Web Audio coordinate space(?), and `pos`is the position in OpenAL space(?), and above that, it is established that transforming between the two needs a change of basis, so these could not be summed just component-wise(?)

jpernst replied:
In listener relative mode, the source position represents a displacement from the listener, expressed in a coordinate system where the listener's forward vector is -Z. The change-of-basis corrects the displacements so they're expressed in world-space instead of listener-space. This is enough for the source direction, since the direction is always relative to the origin anyway, but since the position is a displacement from the listener, we translate it by the listener position to get the final world space coordinates.

I've added a much more detailed comment to explain what's going on here. I'm pretty sure the math is correct, but perhaps more rigorous testing of more complex cases is the only way to be sure.

juj commented on src/library_openal.js:
```diff
@@ -1,1622 +1,4591 @@
+        console.error("alcCaptureOpenDevice() could not create the shared capture AudioContext: " + e);
+#endif
+        // See previously mentioned rationale for ALC_OUT_OF_MEMORY
+        AL.alcErr = 0xA005 /* ALC_OUT_OF_MEM
```
It would be good to document here the specific relationship between `bufQueue` and `audioQueue`

juj replied:
I presume it is not typo to ignore `value` here and pass `null` always unconditionally, but there are no supported parameters to set via this function?

juj replied:
The whole `AL.paramArray` usage could be avoided in this kind of pattern if `AL.setListenerParam` etc. was changed to take in an (optional?) offset to the array being passed, and then the code here could avoid copying these over, but could directly call `AL.setListenerParam("alListeneri3", param, HEAP32, pValues>>2);`, which could then directly read the memory at appropriate location. Not a big deal either way, though just a thought.

juj replied:
In all cases, `buf.frequency = freq;` so this could be hoisted to be a bit less verbose.

juj replied:
In several locations in code, `AL.currentCtx.err` is accessed, but it looks like if no context is current, i.e. `AL.currentCtx` is null, then possibly these might throw. Perhaps these kind of locations should check for `if (AL.currentCtx)` first to avoid leaking exceptions out - ideally as a C API we should never throw toward user.

juj replied:
Why do we expect here that the byte size should be the same? Is that required by the spec? My understanding is that it should be possible to queue different length buffers, and I don't think our implementation cares for them to be the same size?

juj replied:
What is the reason for this zero buffer management?

jpernst replied:
Yeah, some functions have no valid parameters at all, barring possible extensions that add some. In these cases we just delegate to the query function which will handle the error reporting for unknown params. the null value array is a signal to the function that no actual value is expected and to simply report an error.

jpernst replied:
Buffers of different length are infact legal. `bytes` here means bytes per sample. That is a pretty misleading name, so I've changed it to `bytesPerSample`. The spec requires that all queued buffers have the same sample format/channels/frequency, which is what that code is enforcing.

jpernst replied:
The zero-buffer is a particularly strange aspect of the spec. It's legal to use and add to queues, but has odd semantics in some cases. For example, it doesn't impose a format on the queue, since it has no format. The spec is actually a little ambiguous about some of this, but to the best of my reading, if a buffer queue contains only the zero buffer, then it is effectively considered to be empty. I have no idea why they chose to make it this way, and I doubt it will come up in practical use, but I've tried not to editorialize in my implementation.

jpernst replied:
Good catch, I tried to sweep all of these out, but did indeed miss a few. Should be fixed now.

jpernst commented:
Thanks for the review @juj, I'll start looking into these soon. Everything starting at the "bufQueue / audioQueue" one seems to have the same code snippet though, so it's not quite clear what all of them address. I'll try to infer from context for now.

yoanlcq commented:
FWIW, since I suspect it's a GitHub bug, I've just mailed their support along with relevant screenshots.

jpernst commented:
Ok, I've addressed most of the review points I believe. What remains is to decide if the `while (true)` is acceptable or needs an emergency exit, and if the listener relative math is correct. I'm fairly confident on both counts, but it's best to be sure.

jpernst commented:
After reviewing the relative transform code, I did notice one problem, which is that I forget the Up and Forward vectors aren't guaranteed to be orthogonal, which can throw off the transform calculation. I've compensated for this by normalizing and recalculating the basis vectors to ensure an orthonormal basis.

jpernst commented:
Added a comment further explaining the state loop termination conditions, since they're somewhat subtle.

etienne02 commented on tests/openal_capture.c:
```diff
@@ -0,0 +1,298 @@
+#endif
+#include <AL/al.h>
+#include <AL/alc.h>
+#ifdef ASSUME_AL_FLOAT32
```
Is it better to define theses constants in <AL/alext.h> ?
On my linux <AL/ext.h> I have this :
```
#ifndef AL_EXT_float32
#define AL_EXT_float32 1
#define AL_FORMAT_MONO_FLOAT32                   0x10010
#define AL_FORMAT_STEREO_FLOAT32                 0x10011
#endif
```

yoanlcq replied:
Which constants do you mean ?
For `ASSUME_AL_FLOAT32`, it's not the same as `AL_EXT_float32` because the latter only indicates that `<AL/ext.h>` defines the relevant constants, which does not imply that the implementation actually supports the extension.  
For `AL_FORMAT_MONO_FLOAT32` and `AL_FORMAT_STEREO_FLOAT32`, I had decided that is was acceptable because these are only two constants which aren't going to change, ever, and allows us to skip a file inclusion (file which might not actually define these if it is too old (which I believe would be unlikely, but anyway)).

`ASSUME_AL_FLOAT32` represents the hard-coded knowledge that our implementation in Emscripten is guaranteed to support `AL_EXT_float32` at all times.

etienne02 replied:
I mean AL_FORMAT_MONO_FLOAT32 and AL_FORMAT_STEREO_FLOAT32.

yoanlcq replied:
Ok. So yeah, see my rationale above. I do agree that it may "feel" kinda wrong, but since the effects are strictly the same, it even makes the replacement with `#include <AL/ext.h>` not worth doing.  

When more constants are needed, we'll probably want to do it, but right now it's fine as it is.

yoanlcq replied:
Wait, I just realised that Emscripten doesn't have `<AL/ext.h>` (See [the AL folder](https://github.com/jpernst/emscripten/tree/incoming/system/include/AL)), so I think I understand better what you meant.

Am I missing anything ? (@jpernst @juj) We should create one if it's not been done already, and I wouldn't mind doing it.

jpernst replied:
On my system it's `AL/alext.h`. I'm not sure if this is a standard header, since any extension related symbols can be obtained with `alGetEnumValue()` and such, but I'm not opposed to adding it.

etienne02 replied:
Yes its `AL/alext.h` in my system too.

yoanlcq replied:
The `AL/alext.h` header file is not part of the OpenAL 1.1 Core SDK as downloaded on [the official (?) website](https://openal.org/downloads/) (Windows), and neither the spec nor the programming guide mention it (actually they don't mention any header file, unless I missed something).  
Said SDK does include, hovewer, `AL/efx.h`, `AL/efx-creative.h`, `AL/EFX-Util.h` and `AL/xram.h`, which contain preprocessor constants and function pointer typedefs not unlike `AL/alext.h`, which leads me to believe that the latter is a convenience provided by OpenAL-Soft.

Sure, one could say that OpenAL-Soft is the _de facto_ standard, but still not _standard_ in the sense we're interested in.

This further confuses me about what to do. On the one hand, providing an `AL/alext.h` (and friends ? Or not ?) would surely be convenient for Emscripten users, but on the other hand I feel it would be lying (and further propagating the false idea that `AL/alext.h` is standard).  
Furthermore:
1. As @jpernst mentioned, there are already standard ways to fetch extension-related items;
2. Emscripten's OpenAL implementation is likely to implement only a handful of extensions, now and in the future.

I would have added all those files and called it a day if I didn't think this was a big deal - so right now, I'm in favor of keeping things as-is, unless there's a strong reason not to.

jpernst replied:
I don't have a particularly strong preference either way, but if `alext.h` isn't part of the main SDK, then I'm inclined to not bother.

juj replied:
Yeah, looks like we don't have an `AL/alext.h` right now.

> We should create it if it's not been done already, and I wouldn't mind doing it.
> if alext.h isn't part of the main SDK, then I'm inclined to not bother.

> On the one hand, providing an AL/alext.h (and friends ? Or not ?) would surely be convenient for Emscripten users, but on the other hand I feel it would be lying (and further propagating the false idea that AL/alext.h is standard).

On these counts, we can go either way. In Emscripten we are taking a stance of compatibility over standards, and in general we are easy going about adding in semi-nonportable platform specific code, as long as it is not much of a burden that would conflict or confuse. (see library_x11.js and musl compat include directories as an example)

So even if a file `AL/alext.h` is not part of a documented standard, but e.g. OpenAL-Soft specific customization, then we can certainly add that file in, while making a comment in the file noting that it is OpenAL-Soft compatibility specific.

No need to do that in this PR, or if you don't need this functionality, then can leave it out altogether.

However, we would not want to add those files copied verbatim from OpenAL Soft or similar, but probably write them ground up specific to the supported extensions in Emscripten. (and especially if there are some licensing issues)

etienne02 commented on src/library_openal.js:
```diff
@@ -103,898 +193,3990 @@ var LibraryOpenAL = {
+      newCapture.mediaStreamSourceNode = newCapture.audioCtx.createMediaStreamSource(mediaStream);
+
+      var inputChannelCount = 1;
+      switch(newCapture.mediaStreamSourceNode.channelCountMode) {
```
Missing s. It is not `Al.buffer[0]` but `AL.buffers[0]`

etienne02 replied:
Missing s. It should be `AL.buffers[0]` instead of `AL.buffer[0]`

etienne02 commented on src/library_openal.js:
```diff
@@ -103,898 +193,3990 @@ var LibraryOpenAL = {
+      newCapture.mediaStreamSourceNode = newCapture.audioCtx.createMediaStreamSource(mediaStream);
+
+      var inputChannelCount = 1;
+      switch(newCapture.mediaStreamSourceNode.channelCountMode) {
```
Missing s. It should be `AL.buffers[0]` instead of `AL.buffer[0]`

jpernst replied:
Good catch. Fixed.

juj commented:
Ok, I've now got a backlog of other work items cleared, so I should be able to focus on this with better responsiveness - apologies for the wait. Running tests locally, at revision

```
commit 450f18452642d3c15474a092c73f02f30fca3a7a
Author: Jameson Ernst <jameson@jpernst.com>
Date:   Wed Aug 30 08:52:50 2017 -0700

    Fix `AL.buffer` typo
```

I am seeing the following:

 - `python tests/runner.py interactive.test_openal_al_soft_loop_points`

fails with

```
uncaught exception: Assertion failed: state == AL_PLAYING, at: c:/users/clb/appdata/local/temp\emscripten_test_interactive_afurz9\openal_playback.cpp,251,main at jsStackTrace@http://localhost:8888/test.js:1312:13
stackTrace@http://localhost:8888/test.js:1329:12
___assert_fail@http://localhost:8888/test.js:1870:210
_main@http://localhost:8888/test.js:8543:3
asm._main@http://localhost:8888/test.js:15837:10
callMain@http://localhost:8888/test.js:16053:15
doRun@http://localhost:8888/test.js:16117:42
run/<@http://localhost:8888/test.js:16128:7
```

 - `python tests/runner.py interactive.test_openal_animated_looped_distance_playback` has some kind of odd bug that in addition to fading the volume in and out, the pitch of the audio clip changes up and down as well. Reading the test code file, it should not be changing `AL_PITCH`, so I think this pitch change should not be happening in this test?

 - `python tests/runner.py interactive.test_openal_animated_looped_doppler_playback` plays back corrupted/glitchy sound stream. Are you able to reproduce?

 - `python tests/runner.py interactive.test_openal_capture`: There was another PR that landed in between, and this needs a small change from `REPORT_RESULT();` to `REPORT_RESULT(result);` to compile. After that fix, I am getting an issue with the capture not running, and code throwing a stream of exceptions

```
ReferenceError: srcBuf is not defined
	onSuccess/newCapture.scriptProcessorNode.onaudioprocess
```

to web console. I wonder if the capture device I am using might be somehow incompatible, I realize I only have a Oculus Rift headset with its microphone connected, and no other microphones. Or does the issue reproduce for you?

juj commented:
> What remains is to decide if the while (true) is acceptable or needs an emergency exit

Could the loop condition read `for (var i = 0; i < src.bufQueue.length; ++i)`? That way it will be explicit that there's only at most that many iterations.

> if the listener relative math is correct

Re-reading, the code checks out now in my mind as well. A sound source at relative location `(0, 0, -1.0)` should map to world position `lPos + 1.0 * lDirection`, which matches going through the code. Great work adding relative audio support btw, this was missing for a long time.

jpernst commented:
Ok, I can repro all these problems. Odd, there must have been a regression at some point. I'll bisect and see if I can track it down.

jpernst commented:
Ok, I nailed most of those problems, they were all caused by the same bug. The shared param array was accidentally getting cached into some internal fields, so when it was later reused for other param settings, other params would suddenly change, which naturally caused some super wonky breakage.

The capture problem is still an open question, @yoanlcq can probably speak to that better than I.

Anyway, per the loop condition thing: The reason I didn't use that as the loop condition is that if there's a long stall from a blocking operation or something, a looping buffer queue could theoretically have looped many times over by that point, so there's no statically known upper bound to the number of iterations required.

jpernst commented:
Actually, now that I think about it some more, we can probably determine the maximum number of iterations required by calculating the duration of the buffer queue, then dividing the delta time by that. From a strictly semantic standpoint that won't gain is anything, but it may make the loop more clear and further guard against possible non-termination. I'll tinker with it a bit.

yoanlcq commented:
Apologizes for the late reply; I'll see about the capture bug as soon as I can (within the next few days).

jpernst commented:
I've refactored the update loop to remove the unbounded condition, and generally made the update method more clear. This version will probably be faster in some cases as well, since the new control flow lets it skip some unnecessary work in certain cases.

yoanlcq commented:
There we go (hopefully)! For some reason there were two variables (`c` and `srcBuf`) which had vanished (at the same spot), and putting them back just did it.
How this could have happened is beyond me, but hopefully everything's back to normal now.

juj commented:
Perfect, tests all pass now. I am still having issues with microphone recording as well, though my both mic-capable setups are not that great, so I'm not sure. Given that it's a new feature, I'm ok to merge nevertheless, so we can resolve further fixes in the tree.

Thanks for all the work here, and for the great patience to wait for the reviews.
